File: example/python/higgstts_py/tts.py
r"""Unified HiggsTTS API on minitorch.

    tts = HiggsTTS(gguf, tokenizer_json)
    codes = tts.encode_ref(ref_wav_24k)                 # reference audio -> codes [T,8]
    codes = tts.generate(text, codes, ref_text)         # text -> AR codes [T,8]
    pcm   = tts.decode(codes)                           # codes -> 24kHz PCM
    pcm   = tts.synthesize(text, ref_wav, ref_text)     # end to end
"""

import librosa
import numpy as np

import minitorch as mt

from higgstts_py import ar as AR
from higgstts_py import decode as DE
from higgstts_py import model as M
from higgstts_py.weights import HiggsWeights

_TOKENIZER_JSON = r"<local>/reader-app\models\higgs_tts_v3_tokenizer.json"

# prompt special token ids (from higgs_tts.h; GGUF metadata can override)
TOK_TTS = 151667
TOK_REF_TEXT = 151680
TOK_REF_AUDIO = 151679
TOK_TEXT = 151672
TOK_AUDIO = 151670
AUDIO_PLACEHOLDER = -100
N_CB = 8

_SEM_RATE = 16000
_AC_RATE = 24000
_HOP = int(np.prod(M.AC_STRIDES))
_FE_K = [10, 3, 3, 3, 3, 2, 2]
_FE_S = [5, 2, 2, 2, 2, 2, 2]


def _fe_len(L):
    for K, s in zip(_FE_K, _FE_S):
        L = (L - K) // s + 1
    return L


def _ac_len(L):
    def cl(L, K, s, p):
        return (L + 2 * p - (K - 1) - 1) // s + 1
    L = cl(L, 7, 1, 3)
    for K, s, p in zip([16, 10, 8, 4, 6], M.AC_STRIDES, M.AC_PADS):
        L = cl(L, K, s, p)
    return cl(L, 3, 1, 1)


class HiggsTTS:
    def __init__(self, gguf_path, tokenizer_json=_TOKENIZER_JSON, device=None, runtime=None):
        self.rt = runtime or mt.Runtime()
        self.dev = device or self.rt.gpu()
        self.w = HiggsWeights(gguf_path, self.dev, backbone=True)
        M.build_pce_weight(self.w)
        self.tok = None
        if tokenizer_json:
            from tokenizers import Tokenizer
            self.tok = Tokenizer.from_file(tokenizer_json)

    # ---- 1. reference audio -> codes ----
    def encode_ref(self, waveform, sample_rate=24000):
        wav = np.asarray(waveform, dtype=np.float32).reshape(-1)
        if sample_rate != _AC_RATE:
            wav = librosa.resample(wav, orig_sr=sample_rate, target_sr=_AC_RATE).astype(np.float32)
        sem = np.pad(librosa.resample(wav, orig_sr=_AC_RATE, target_sr=_SEM_RATE).astype(np.float32),
                     (160, 160))
        t_sem = _fe_len(len(sem))
        t_ac = _ac_len(len(wav))
        pad = t_sem - t_ac
        pcm = np.pad(wav, ((pad // 2) * _HOP, (pad - pad // 2) * _HOP)) if pad > 0 else wav

        g = mt.Graph(self.rt, self.dev)
        g.enter()
        xs = g.input([sem.shape[0], 1], sem.tobytes())
        xa = g.input([pcm.shape[0], 1], pcm.tobytes())
        ids, _ = M.prefill(self.w, xs, xa)
        id_c = [mt.contiguous(i) for i in ids]
        for t in id_c:
            t.mark_output()
        g.exit()
        t = id_c[0].shape[0]
        codes = np.zeros((t, N_CB), dtype=np.int32)
        for q, node in enumerate(id_c):
            codes[:, q] = np.frombuffer(node.to_bytes(), dtype=np.int32).reshape(-1)
        return codes

    # ---- prompt ----
    def build_prompt(self, text, ref_text, num_ref_tokens):
        if self.tok is None:
            raise RuntimeError("HiggsTTS: no tokenizer loaded")
        p = [TOK_TTS]
        if ref_text:
            p.append(TOK_REF_TEXT)
            p += self.tok.encode(ref_text, add_special_tokens=False).ids
        p.append(TOK_REF_AUDIO)
        p += [AUDIO_PLACEHOLDER] * num_ref_tokens
        p.append(TOK_TEXT)
        p += self.tok.encode(text, add_special_tokens=False).ids
        p.append(TOK_AUDIO)
        return np.asarray(p, dtype=np.int32)

    # ---- 2. text + ref codes -> AR codes ----
    def generate(self, text, ref_codes, ref_text="", temperature=0.9, seed=42, topk=50,
                 max_steps=400):
        num_ref = np.asarray(ref_codes).shape[0] + (N_CB - 1)   # + delay pattern
        prompt = self.build_prompt(text, ref_text, num_ref)
        return AR.ar_generate(self.rt, self.w, prompt, np.asarray(ref_codes, dtype=np.int32),
                              temperature=temperature, seed=seed, max_steps=max_steps, topk=topk)

    # ---- 3. codes -> PCM ----
    def decode(self, codes):
        return DE.dac_decode(self.rt, self.w, np.asarray(codes, dtype=np.int32))

    # ---- end to end ----
    def synthesize(self, text, ref_wav, ref_text="", sample_rate=24000, temperature=0.9,
                   seed=42, topk=50, max_steps=400):
        ref_codes = self.encode_ref(ref_wav, sample_rate)
        codes = self.generate(text, ref_codes, ref_text, temperature, seed, topk, max_steps)
        return self.decode(codes)

File: example/CSharp/HiggsTtsSharp/Tts.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using HiggsTts;
using VulkanTorch;

namespace HiggsTtsSharp;

/// <summary>Synthesis options (mirror the HiggsTTS CLI / higgstts_vt defaults).</summary>
public sealed class HiggsOptions
{
    public float Temperature { get; set; } = 0.9f;
    public int TopK { get; set; } = 50;
    public int Seed { get; set; } = 42;
    /// <summary>AR step budget; 0 = predict from text length (12 * text tokens + 200).</summary>
    public int MaxSteps { get; set; }
    /// <summary>Replay the captured decode graph instead of rebuilding it each step.</summary>
    public bool GraphCache { get; set; } = true;
    /// <summary>Optional style tag (e.g. <c>&lt;|style:whispering|&gt;</c>) prepended to the text.</summary>
    public string? StyleTag { get; set; }

    /// <summary>Streaming: frames of context each decode window carries on both sides of
    /// what it emits. Latency vs. closeness to a one-shot decode — see
    /// <see cref="DacStream.DefaultLookahead"/>.</summary>
    public int StreamLookaheadFrames { get; set; } = DacStream.DefaultLookahead;
    /// <summary>Streaming: frames per emitted chunk. 8 → 320 ms of audio per chunk.</summary>
    public int StreamStepFrames { get; set; } = 8;
    /// <summary>Streaming: samples of trailing silence that end the stream early. 0 or
    /// negative disables the rule (used when comparing against a one-shot decode, which
    /// has no such cut-off).</summary>
    public int SilenceStopSamples { get; set; } = 4 * Tts.SampleRate;
}

/// <summary>
/// HiggsTTS synthesis facade: reference audio → <see cref="ReferenceVoice"/>, then
/// text + voice → 24 kHz PCM. Self-contained (owns its Runtime / device / weights).
/// Not thread-safe — serialize calls on one instance.
/// </summary>
public sealed class Tts : IDisposable
{
    public const int SampleRate = 24000;

    // prompt special token ids (from higgs_tts.h / higgstts_vt/tts.py)
    private const int TokTts = 151667, TokRefText = 151680, TokRefAudio = 151679;
    private const int TokText = 151672, TokAudio = 151670, AudioPlaceholder = -100;

    private readonly Runtime _rt;
    private readonly Device _dev;
    private readonly HiggsWeights _w;
    private HiggsTokenizer? _tok;
    private bool _disposed;

    public Tts(string modelGgufPath, string? tokenizerJsonPath = null)
    {
        _rt = new Runtime();
        _dev = _rt.Gpu();
        _w = new HiggsWeights(modelGgufPath, _dev,
            HiggsTts.HiggsWeights.PrefillPrefixes.Concat(HiggsTts.HiggsWeights.BackbonePrefixes).ToArray(),
            4UL << 30);
        EncodeRef.BuildPceWeight(_w);
        // The GGUF carries the vocab+merges, so a separate tokenizer.json is optional.
        _tok = !string.IsNullOrWhiteSpace(tokenizerJsonPath)
            ? new HiggsTokenizer(tokenizerJsonPath)
            : HiggsTokenizer.FromGguf(modelGgufPath);
    }

    /// <summary>Override the BPE tokenizer with an HF tokenizer.json (empty path = keep the GGUF-built one).</summary>
    public bool SetTokenizer(string tokenizerJsonPath)
    {
        if (string.IsNullOrWhiteSpace(tokenizerJsonPath)) return true;
        _tok = new HiggsTokenizer(tokenizerJsonPath);
        return true;
    }

    public bool HasTokenizer => _tok != null;

    /// <summary>Encode reference audio (any sample rate) → a reusable <see cref="ReferenceVoice"/>.
    ///
    /// This is the one part of a request that does not depend on the text, and it is not
    /// cheap (~125 ms for a 15 s reference), so a server should call <see cref="Warmup"/>
    /// once per voice at startup instead of letting the first request pay for it.</summary>
    public ReferenceVoice EncodeReference(float[] audio, int sampleRate, string? refText = null)
    {
        var codes = EncodeRef.Run(_rt, _dev, _w, audio, sampleRate);   // int[T, 8]; resamples internally
        int t = codes.GetLength(0);
        var flat = new int[t * Ar.NCb];
        for (int i = 0; i < t; i++)
            for (int c = 0; c < Ar.NCb; c++) flat[i * Ar.NCb + c] = codes[i, c];
        return new ReferenceVoice(flat, refText);
    }

    /// <summary>Text + voice → raw RVQ codes (flat [T*8], row-major).</summary>
    public int[] GenerateCodes(string text, ReferenceVoice voice, HiggsOptions? options = null)
    {
        options ??= new HiggsOptions();
        if (_tok == null) throw new InvalidOperationException("no tokenizer; call SetTokenizer first");
        if (!string.IsNullOrEmpty(options.StyleTag)) text = options.StyleTag + text;

        var prompt = BuildPrompt(_tok, text, voice.RefText, voice.Rows + Ar.NCb - 1);
        return Ar.Generate(_rt, _dev, _w, prompt, voice.Codes, voice.Rows, out _,
                           options.Temperature, options.Seed, options.MaxSteps, options.TopK, options.GraphCache);
    }

    /// <summary>Raw RVQ codes → 24 kHz float32 PCM.</summary>
    public float[] Decode(int[] codes) => DacDecoder.Decode(_rt, _dev, _w, codes, codes.Length / Ar.NCb);

    /// <summary>Text + voice → 24 kHz float32 PCM.</summary>
    public float[] Synthesize(string text, ReferenceVoice voice, HiggsOptions? options = null)
        => Decode(GenerateCodes(text, voice, options));

    /// <summary>Same as <see cref="GenerateCodes"/>, but each completed code frame is
    /// passed to <paramref name="onFrame"/> as soon as the AR produces it. Returning false
    /// stops generation early (client gone). See <see cref="Ar.GenerateStreaming"/>.</summary>
    public void GenerateCodesStreaming(string text, ReferenceVoice voice, HiggsOptions? options,
                                       Func<int[], bool> onFrame)
    {
        options ??= new HiggsOptions();
        if (_tok == null) throw new InvalidOperationException("no tokenizer; call SetTokenizer first");
        if (!string.IsNullOrEmpty(options.StyleTag)) text = options.StyleTag + text;

        var prompt = BuildPrompt(_tok, text, voice.RefText, voice.Rows + Ar.NCb - 1);
        Ar.GenerateStreaming(_rt, _dev, _w, prompt, voice.Codes, voice.Rows,
                             options.Temperature, options.Seed, options.MaxSteps, options.TopK,
                             options.GraphCache, onFrame);
    }

    /// <summary>Text + voice → PCM in chunks as it is produced, instead of one array at
    /// the end. Runs the AR and the DAC on this thread in lockstep (the reference server
    /// decodes inside its AR callback too) — nothing here is concurrent.
    ///
    /// <paramref name="onPcm"/> gets ~<see cref="HiggsOptions.StreamStepFrames"/> frames'
    /// worth of 24 kHz float PCM per call. Returning false stops generation (client
    /// disconnected). Generation also stops on its own after roughly four seconds of
    /// trailing silence, which the AR sometimes produces without ever emitting EOC.
    ///
    /// Returns how many frames' audio was actually handed over, and whether the silence
    /// rule is what ended it.</summary>
    public (int Frames, bool StoppedForSilence) SynthesizeStreaming(
        string text, ReferenceVoice voice, HiggsOptions? options,
        Func<ReadOnlyMemory<float>, bool> onPcm)
        => SynthesizeStreamingCodes(text, voice, options, null, onPcm);

    /// <summary>As <see cref="SynthesizeStreaming"/>, but also hands every generated code
    /// frame to <paramref name="onFrame"/> — so a caller can keep the codes that produced
    /// the audio it just heard, instead of having to reproduce the run.</summary>
    public (int Frames, bool StoppedForSilence) SynthesizeStreamingCodes(
        string text, ReferenceVoice voice, HiggsOptions? options,
        Func<int[], bool>? onFrame, Func<ReadOnlyMemory<float>, bool> onPcm)
    {
        var dac = NewStream(options, onPcm);
        GenerateCodesStreaming(text, voice, options, frame =>
            (onFrame == null || onFrame(frame)) && dac.Push(frame, onPcm));
        FinishStream(dac, onPcm);
        return (dac.DecodedFrames, dac.StoppedForSilence);
    }

    /// <summary>Decode an already-generated code block in streaming windows rather than
    /// in one pass. Same audio as <see cref="Decode"/> apart from the decode-window
    /// boundaries; exists so a fixed code block can be run down both paths and compared
    /// (<c>decode --codes ... --stream</c>).</summary>
    public (int Frames, bool StoppedForSilence) DecodeStreaming(
        int[] codes, HiggsOptions? options, Func<ReadOnlyMemory<float>, bool> onPcm)
    {
        var dac = NewStream(options, onPcm);
        int n = codes.Length / Ar.NCb;
        for (int t = 0; t < n; t++)
        {
            var frame = new int[Ar.NCb];
            Array.Copy(codes, t * Ar.NCb, frame, 0, Ar.NCb);
            if (!dac.Push(frame, onPcm)) break;
        }
        FinishStream(dac, onPcm);
        return (dac.DecodedFrames, dac.StoppedForSilence);
    }

    private DacStream NewStream(HiggsOptions? options,
                                Func<ReadOnlyMemory<float>, bool> onPcm)
    {
        var dac = new DacStream(_rt, _dev, _w);
        if (options != null)
        {
            dac.Lookahead = options.StreamLookaheadFrames;
            dac.Step = options.StreamStepFrames;
            dac.SilenceStopSamples = options.SilenceStopSamples;
        }
        return dac;
    }

    private static void FinishStream(DacStream dac, Func<ReadOnlyMemory<float>, bool> onPcm)
    {
        // Nothing left to hold silence back for, so a stop here would only drop the tail.
        if (!dac.StoppedForSilence) dac.Flush(onPcm);
    }

    private static int[] BuildPrompt(HiggsTokenizer tok, string text, string? refText, int numRef)
    {
        var p = new List<int> { TokTts };
        if (!string.IsNullOrEmpty(refText))
        {
            p.Add(TokRefText);
            p.AddRange(tok.Encode(refText));
        }
        p.Add(TokRefAudio);
        for (int i = 0; i < numRef; i++) p.Add(AudioPlaceholder);
        p.Add(TokText);
        if (!string.IsNullOrEmpty(text)) p.AddRange(tok.Encode(text));
        p.Add(TokAudio);
        return p.ToArray();
    }

    public string? BackendName => _rt.Name;

    /// <summary>Number of weights loaded (introspection).</summary>
    public int TensorCount => _w.Count;

    /// <summary>Tokenize text with the loaded BPE (introspection; throws if none set).</summary>
    public int[] EncodeText(string text)
        => _tok != null ? _tok.Encode(text) : throw new InvalidOperationException("no tokenizer");

    public void Dispose()
    {
        if (_disposed) return;
        _w.Dispose();
        _rt.Dispose();
        _disposed = true;
    }
}
